sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -872,6 +886,25 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
       pushDownPredicate(filter, u.child) { predicate =>
         u.withNewChildren(Seq(Filter(predicate, u.child)))
       }
+
+    case filter @ Filter(condition, watermark: EventTimeWatermark) =>
```

Comment:
For filter, the logic has a subtle additional condition. We don't want to push down filters on the watermark attribute, because:
* they'll be at least as expensive to evaluate as the watermark operator
* partition pruning shouldn't apply since there won't be useful partitions on an event time

For project, I don't see a rule for UnaryNode anywhere. I might have missed it.

For limit, I actually removed the rule for EventTimeWatermark that I originally added, since it does drop rows in some situations.

So I don't think that making EventTimeWatermark subclass UnaryNode would avoid any of the code in this PR. I agree it seems appropriate, but it also seems orthogonal.